flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartition.java

```diff
@@ -375,4 +398,33 @@ void onConsumedSubpartition(int subpartitionIndex) {
 	private void checkInProduceState() throws IllegalStateException {
 		checkState(!isFinished, "Partition already finished.");
 	}
+
+	/**
+	 * Check whether all subpartitions' backlogs are less than the limitation of max backlogs, and make this partition
+	 * available again if yes.
+	 */
+	@GuardedBy("buffers")
```

Comment:
That's not true. First of all this class can not relay on an implementation detail of the `PipelinedSubpartition` (which is only one of two implementations of `ResultSubpartition` class).

Secondly each `PipelinedSubpartition` has it's own `buffers` field, so the updates here to `unavailableSubpartitionsCnt` are complete not synchronized. Some simple solution might be to change `unavailableSubpartitionsCnt` to an `AtomicLong`, but let me think about alternative solutions.

Reply:
Another solution comes to my mind is BitSet structure and each subpartition take care of its own single bit. We can easily construct such a structure by using `long[]` (new long[1024] can be used by 64 * 1024 subpartitions). And instead of using a counter, we can use bit operations to check whether all long values in `long[]` are 0.

Reply:
I would expect the smallest atomic writes/reads would be bytes, not bits. With longs, threads performing
```
long l = bitSet[i];
l |= 1 << n;
bitSet[i] = l;
```
would be overriding bits from other threads sharing the same long. So you would need single byte per subpartition.